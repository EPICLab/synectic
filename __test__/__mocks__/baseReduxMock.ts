import { DateTime } from 'luxon';
import { v4 } from 'uuid';
import parsePath from 'parse-path';

import { Canvas, Stack, Card, Filetype, Metafile, Repository, Metadir } from '../../src/types';
import { createStore } from 'redux';
import { rootReducer } from '../../src/store/root';

export const getMockStore = () => {
  type initStateT = {
    canvas: { [id: string]: Canvas };
    stacks: { [id: string]: Stack };
    cards: { [id: string]: Card };
    filetypes: { [id: string]: Filetype };
    metafiles: { [id: string]: Metafile };
    repos: { [id: string]: Repository };
    metadirs: { [id: string]: Metadir };
  }
  const initialState: initStateT = {
    canvas: {},
    stacks: {
      33: {
        id: '33',
        name: 'sample stack',
        created: DateTime.fromISO('2019-01-21T08:14:52.181-08:00'),
        modified: DateTime.fromISO('2019-11-19T19:22:47.572-08:00'),
        note: '',
        cards: [],
        left: 5,
        top: 5
      }
    },
    cards: {
      14: {
        id: '14',
        name: 'test.js',
        type: 'Editor',
        related: ['243'],
        created: DateTime.fromISO('2019-01-21T08:14:52.181-08:00'),
        modified: DateTime.fromISO('2019-11-19T19:22:47.572-08:00'),
        captured: false,
        left: 10,
        top: 10
      },
      29: {
        id: '29',
        name: 'example.ts',
        type: 'Editor',
        related: ['199'],
        created: DateTime.fromISO('2019-01-21T08:14:52.181-08:00'),
        modified: DateTime.fromISO('2019-11-19T19:22:47.572-08:00'),
        captured: false,
        left: 10,
        top: 10
      }
    },
    filetypes: {
      91: {
        id: '91',
        filetype: 'JavaScript',
        handler: 'Editor',
        extensions: ['js', 'jsm']
      },
      45: {
        id: '45',
        filetype: 'TypeScript',
        handler: 'Editor',
        extensions: ['ts', 'typescript', 'str']
      }
    },
    metafiles: {
      199: {
        id: '199',
        name: 'test.js',
        path: 'withchildren/test.js',
        modified: DateTime.fromISO('2019-11-19T19:19:47.572-08:00')
      },
      243: {
        id: '243',
        name: 'example.ts',
        modified: DateTime.fromISO('2015-06-19T19:10:47.572-08:00')
      }
    },
    repos: {
      13: {
        id: '13',
        name: 'test/repo',
        corsProxy: new URL('http://www.random_proxy.edu'),
        url: parsePath('http://www.random_proxy.edu'),
        refs: ['master'],
        oauth: 'github',
        username: 'sam',
        password: 'pass123',
        token: '934394304234231'
      }
    },
    metadirs: {
      99: {
        id: '99',
        name: 'testdir',
        path: 'withchildren/testdir',
        expanded: false,
        containsDir: [],
        containsFile: []
      },
      24: {
        id: '24',
        name: 'withchildren',
        path: 'withchildren',
        expanded: false,
        containsDir: ["withchildren/testdir"],
        containsFile: ['withchildren/test']
      }
    }
  };
  return createStore(rootReducer, initialState);
};

export const getCanvasProps = (): Canvas => {
  return {
    id: v4(),
    created: DateTime.fromISO('2019-11-19T19:22:47.572-08:00'),
    repos: ['13'],
    cards: ['29', '14'],
    stacks: ['33']
  };
};