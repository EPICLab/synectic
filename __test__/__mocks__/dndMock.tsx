/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable react-hooks/rules-of-hooks */
/* eslint-disable import/named */

/**
 * The react-dnd library has an issue with Enzyme tests that interact with the DragDropManager on
 * context-aware components, per https://github.com/react-dnd/react-dnd/issues/1506. This solution
 * was created by JKillian (https://github.com/JKillian) and published here: 
 * https://github.com/react-dnd/react-dnd/pull/1570
 */
import React from 'react';
import TestBackendImpl from 'react-dnd-test-backend';
import { DndProvider, DndContext } from 'react-dnd';
import { DragDropManager } from 'dnd-core';

interface RefType {
  getManager: () => DragDropManager | undefined;
}

export function wrapInTestContext(DecoratedComponent: any): any {
  const forwardRefFunc = (props: any, ref: React.Ref<RefType>) => {
    const dragDropManager = React.useRef<any>(undefined);

    React.useImperativeHandle(ref, () => ({
      getManager: () => dragDropManager.current,
    }));

    return (
      <DndProvider backend={TestBackendImpl}>
        <DndContext.Consumer>
          {ctx => {
            dragDropManager.current = ctx.dragDropManager;
            return null;
          }}
        </DndContext.Consumer>
        <DecoratedComponent {...props} />
      </DndProvider>
    );
  };
  forwardRefFunc.displayName = 'TestContextWrapper';

  return React.forwardRef(forwardRefFunc);
}