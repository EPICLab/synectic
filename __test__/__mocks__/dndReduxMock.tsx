/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable react-hooks/rules-of-hooks */
/* eslint-disable import/named */

/**
 * The react-dnd library has an issue with Enzyme tests that interact with the DragDropManager on
 * context-aware components, per https://github.com/react-dnd/react-dnd/issues/1506. This solution
 * was created by JKillian (https://github.com/JKillian) and published here: 
 * https://github.com/react-dnd/react-dnd/pull/1570
 */
import React from 'react';
import TestBackendImpl from 'react-dnd-test-backend';
import { DndProvider, DndContext } from 'react-dnd';
import { DragDropManager } from 'dnd-core';
import { Store, AnyAction } from 'redux';
import { Provider } from 'react-redux';

interface RefType {
  getManager: () => DragDropManager | undefined;
}

export function wrapInTestContext(DecoratedComponent: any, store: Store<any, AnyAction>): any {
  const forwardRefFunc = (props: any, ref: React.Ref<RefType>) => {
    const dragDropManager = React.useRef<any>(undefined);

    React.useImperativeHandle(ref, () => ({
      getManager: () => dragDropManager.current,
    }));

    return (
      <Provider store={store}>
        <DndProvider backend={TestBackendImpl}>
          <DndContext.Consumer>
            {ctx => {
              dragDropManager.current = ctx.dragDropManager;
              return null;
            }}
          </DndContext.Consumer>
          <DecoratedComponent {...props} />
        </DndProvider>
      </Provider>
    );
  };
  forwardRefFunc.displayName = 'TestContextWrapper';

  return React.forwardRef(forwardRefFunc);
}