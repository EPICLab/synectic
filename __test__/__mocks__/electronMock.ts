const sampleFilePaths = ['./__test__/__mocks__/fileMock.js', './__test__/__mocks__/reduxStoreMock.ts'];
const sampleFilePath = './__test__/__mocks__/fileMock.js';

export const remote = {
  dialog: {
    // replace the showOpenDialog and showSaveDialog functions with spies which return a valid value
    showOpenDialog: jest.fn().mockReturnValue(new Promise((resolve) => resolve({ canceled: false, filePaths: sampleFilePaths }))),
    showSaveDialog: jest.fn().mockReturnValue(new Promise((resolve) => resolve({ canceled: false, filePath: sampleFilePath })))
  }
}