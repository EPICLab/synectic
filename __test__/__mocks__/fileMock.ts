const rand: number = Math.floor(Math.random() * 6) + 1;