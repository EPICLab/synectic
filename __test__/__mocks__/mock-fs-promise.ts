import { dir } from 'tmp-promise';
import * as fs from 'fs-extra';
import * as path from 'path';
import { MockedRepository, mockGit, MockInstanceEnhanced } from './mock-git-promise';

type Encoding = 'ascii' | 'base64' | 'hex' | 'ucs2' | 'ucs-2' | 'utf16le' | 'utf-16le' | 'utf8' | 'utf-8' | 'binary' | 'latin1';
type Octal = 0o0 | 0o1 | 0o2 | 0o3 | 0o4 | 0o5 | 0o6 | 0o7;
type Mode = `${Octal}${Octal}${Octal}`;
type Flag = 'a' | 'ax' | 'a+' | 'ax+' | 'as' | 'as+' | 'r' | 'r+' | 'rs+' | 'w' | 'wx' | 'w+' | 'wx+';

export type MockInstance = {
    getRoot(): string;
    addItem(filename: string, item: DirectoryItem): Promise<void>;
    reset(): void;
}

export type DirectoryItem =
    | string
    | Buffer
    | File
    | Directory
    | SymbolicLink
    | DirectoryItems;

export type DirectoryItems = {
    [name: string]: DirectoryItem;
}

export type File = {
    content: string | Buffer;
    mode?: Mode | undefined;
    encoding?: Encoding | undefined;
    flag?: Flag | undefined;
    uid?: number | undefined;
    gid?: number | undefined;
    atime?: Date | undefined;
    mtime?: Date | undefined;
}

export type Directory = {
    items: DirectoryItems | undefined;
    mode?: Mode | undefined;
    encoding?: Encoding | undefined;
    flag?: Flag | undefined;
    uid?: number | undefined;
    gid?: number | undefined;
    atime?: Date | undefined;
    mtime?: Date | undefined;
    birthtime?: Date | undefined;
}

export type SymbolicLink = {
    path: string;
    mode?: Mode | undefined;
    encoding?: Encoding | undefined;
    flag?: Flag | undefined;
    uid?: number | undefined;
    gid?: number | undefined;
    atime?: Date | undefined;
    mtime?: Date | undefined;
    birthtime?: Date | undefined;
}

export async function mock(config: DirectoryItems, gitRepo: MockedRepository): Promise<MockInstanceEnhanced>;
export async function mock(config: DirectoryItems): Promise<MockInstance>;
export async function mock(config: DirectoryItems, gitRepo?: MockedRepository | undefined): Promise<MockInstance | MockInstanceEnhanced> {
    const root = await dir({ unsafeCleanup: true });
    const cwd = process.cwd();

    process.chdir(root.path);
    for (const entry of Object.entries(config)) await create(entry);

    const instance = {
        getRoot: () => root.path,
        addItem: (filename: string, item: DirectoryItem) => create([filename, item]),
        reset: () => {
            root.cleanup();
            process.chdir(cwd);
        }
    };
    if (gitRepo) return mockGit(instance, gitRepo);
    else return instance;
}

export const create = async ([filename, config]: [string, DirectoryItem]): Promise<void> => {
    if (isContent(config)) {
        await writeFile(filename, config);
    } else if (isFile(config)) {
        const options: fs.WriteFileOptions = {
            mode: config.mode ? parseInt(config.mode, 8) : undefined,
            encoding: config.encoding,
            flag: config.flag
        };
        await writeFile(filename, config.content, options);

        if (config.atime && config.mtime) await fs.utimes(filename, config.atime, config.mtime);
        if (config.atime && !config.mtime) await fs.utimes(filename, config.atime, new Date(1));
        if (!config.atime && config.mtime) await fs.utimes(filename, new Date(1), config.mtime);
    } else if (isDirectory(config)) {
        await writeDir(filename, config.mode ? parseInt(config.mode, 8) : undefined);
    } else if (isSymlink(config)) {
        await writeSymlink(filename, config.path);
    } else if (isDirectoryItems(config)) {
        const items = Object.entries(config);
        if (items.length > 0) {
            for (const [subfile, subconfig] of items) await create([path.join(filename, subfile), subconfig]);
        } else {
            await writeDir(filename);
        }
    }
}

const isContent = (config: DirectoryItem): config is string | Buffer => {
    return (typeof config === 'string' || config instanceof Buffer);
}

export const file = (config?: File): File => {
    return { content: config ? config.content : '', ...config }
}

const isFile = (config: DirectoryItem): config is File => {
    return (config as File).content !== undefined;
}

export const directory = (config?: Directory): Directory => {
    return { items: config ? config.items : undefined, ...config }
}

const isDirectory = (config: DirectoryItem): config is Directory => {
    return (config as Directory).items !== undefined;
}

export const symlink = (config?: SymbolicLink): SymbolicLink => {
    return { path: config ? config.path : '', ...config }
}

const isSymlink = (config: DirectoryItem): config is SymbolicLink => {
    return (config as SymbolicLink).path !== undefined;
}

const isDirectoryItems = (config: DirectoryItem): config is DirectoryItems => {
    return Object.keys(config).length !== undefined;
}


const writeFile = async (filepath: fs.PathLike, content: string | Buffer, options?: fs.WriteFileOptions) => {
    try {
        await fs.ensureFile(filepath.toString());
        await fs.outputFile(filepath.toString(), content, options);
    } catch (error) {
        console.error(error);
    }
}

const writeDir = async (dirpath: fs.PathLike, options?: number | { mode: number }) => {
    try {
        await fs.ensureDir(dirpath.toString(), options);
    } catch (error) {
        console.error(error);
    }
}

const writeSymlink = async (srcPath: fs.PathLike, destPath: fs.PathLike) => {
    try {
        await fs.ensureSymlink(srcPath.toString(), destPath.toString());
    } catch (error) {
        console.error(error);
    }
}