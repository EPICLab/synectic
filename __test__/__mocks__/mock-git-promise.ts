import * as fs from 'fs-extra';
import * as path from 'path';
import * as git from 'isomorphic-git';
import * as casual from 'casual';
import sha1 from 'sha1';
import { MockInstance } from './mock-fs-promise';
import { readDirAsyncDepth, readFileAsync, writeFileAsync } from '../../src/containers/io';

/**
 * Accept some meta-information about the repository, and then a path to an existing .git directory
 * in order to create a consolidated packfile version of the git repository, which can be unpacked
 * for use as a mocked git repository inside of a mock-fs-promise instance.
 * 
 * Design principles for this Git repository mocking library (extending from `mock-fs-promise`):
 *   (1) Create fully functional local Git repositories
 *   (2) Deterministic repositories for simplified testing
 *   (3) Quick clean-up process for removing all filesystem entries
 *   (4) Support for advanced structures (i.e. linked worktrees)
 *   (5) Mocked Git server for responding to Git requests (i.e. clone, fetch, etc.)
 */

type SHA1 = ReturnType<typeof sha1>;

// Descriminated union type for emulating a `mutually exclusive or` (XOR) operation between parameter types
// Ref: https://github.com/microsoft/TypeScript/issues/14094#issuecomment-344768076
type IncludeOrExcludeList = { include: fs.PathLike[], exclude?: never } | { include?: never, exclude: fs.PathLike[] };

export type MockInstanceEnhanced = MockInstance & {
    getRepoRoot(): string;
    getCommits(): SHA1[];
    stage(filepath: fs.PathLike): Promise<void>;
    commit(commit: MockedCommit): Promise<string>;
}

type MockedCommit = {
    oid: SHA1,
    message: string,
    author: Partial<{
        name: string,
        email: string,
        timestamp: number,
        timezoneOffset: number
    }>,
    committer: Partial<{
        name: string,
        email: string,
        timestamp: number,
        timezoneOffset: number
    }>,
    files: 'all' | IncludeOrExcludeList
}

type MockedBranch = {
    name: string,
    base: string,
    ahead: MockedCommit[],
    behind?: number
}

export type MockedRepository = {
    config?: Partial<{
        'user.name': string,
        'user.email': string
    }>,
    url?: string,
    default: string,
    branches: MockedBranch[]
}

export const mockGit = async (instance: MockInstance, repo: MockedRepository): Promise<MockInstanceEnhanced> => {
    const dir = instance.getRoot();
    const gitdir = path.resolve(dir, '.git');
    const commits: SHA1[] = [];

    // check if a git repository already exists
    if (!(await fs.pathExists(gitdir))) {
        await git.init({ fs, dir, defaultBranch: repo.default });
        await git.setConfig({
            fs,
            dir,
            path: 'user.name',
            value: repo.config['user.name'] ? repo.config['user.name'] : casual.full_name
        });
        await git.setConfig({
            fs,
            dir,
            path: 'user.email',
            value: repo.config['user.email'] ? repo.config['user.email'] : casual.email
        });
        if (repo.url) {
            await git.addRemote({
                fs,
                dir,
                remote: 'origin',
                url: repo.url
            })
        }
        for (const branch of repo.branches) {
            const newCommits = await writeBranch(dir, branch);
            newCommits.map(commit => commits.push(commit));
        }
    }

    const stage = (filepath: fs.PathLike): Promise<void> => {
        return git.add({
            fs,
            dir: dir,
            filepath: filepath.toString()
        });
    }

    const commit = (commit: MockedCommit): Promise<string> => {
        return git.commit({
            fs,
            dir,
            message: commit.message,
            author: commit.author,
            committer: commit.author
        });
    }

    return {
        ...instance,
        getRepoRoot: () => gitdir,
        getCommits: () => commits,
        stage,
        commit
    };
}

const gitAddMany = async (dir: fs.PathLike, files: fs.PathLike[]) =>
    Promise.all(files.map(file => git.add({
        fs,
        dir: dir.toString(),
        filepath: file.toString()
    })));

const writeBranch = async (dir: string, branch: MockedBranch): Promise<SHA1[]> => {
    const commits: SHA1[] = [];

    // checkout the base branch into the working tree, if needed
    if ((await git.listBranches({ fs, dir })).length > 0) await git.checkout({
        fs,
        dir,
        ref: branch.base,
    });
    // branch off to the new branch
    await git.branch({
        fs,
        dir,
        ref: branch.name
    });
    // reset HEAD to point to previous commits, if needed
    if (branch.behind) {
        await gitReset(dir, `HEAD~${branch.behind}`, branch.name, true);
    }

    // add all branch-specific commits
    for (const [index, commit] of branch.ahead.entries()) {
        const allFiles = (await readDirAsyncDepth(dir)).map(file => path.relative(dir, file)).filter(filepath => filepath.length > 0);

        if (commit.files === 'all') {
            if (index > 0) await randomFileUpdate(allFiles);
            await gitAddMany(dir, allFiles);
        } else if (commit.files.include) {
            if (index > 0) await randomFileUpdate(commit.files.include);
            await gitAddMany(dir, commit.files.include);
        } else if (commit.files.exclude) {
            const stringifiedExcludes = commit.files.exclude.map(file => file.toString());
            const includedFiles = allFiles.filter(file => !stringifiedExcludes.includes(file));
            if (index > 0) await randomFileUpdate(includedFiles);
            await gitAddMany(dir, includedFiles);
        }

        commits.push(await git.commit({
            fs,
            dir,
            message: commit.message,
            author: commit.author,
            committer: commit.committer
        }));
    }
    return commits;
}

// Copied from `jcubic`: https://github.com/jcubic/git/blob/gh-pages/js/main.js#L1915
// Same as `git reset --soft` or `git reset --hard` for moving back the current HEAD pointer
// on a git branch.
const gitReset = async (dir: string, ref: string, branch: string, hard = false) => {
    const re = /^HEAD~([0-9]+)$/;
    const match = ref.match(re);
    if (match) {
        const count = +match[1];
        const commits = await git.log({ dir, fs, depth: count + 1 });
        return new Promise<void>((resolve, reject) => {
            if (commits.length < count + 1) {
                return reject('Not enough commits');
            }
            const commit = commits.pop().oid;
            fs.writeFile(`${dir}/.git/refs/heads/${branch}`, commit + '\n', (err) => {
                if (err) {
                    return reject(err);
                }
                if (!hard) {
                    resolve();
                } else {
                    // clear the index (if any)
                    fs.unlink(`${dir}/.git/index`, (err) => {
                        if (err) {
                            return reject(err);
                        }
                        // checkout the branch into the working tree
                        git.checkout({ fs, dir, ref: branch }).then(resolve);
                    })
                }
            })
        })
    }
    return Promise.reject(`Wrong ref ${ref}`);
}

// given a set of filepaths, either targeted or randomly select file and update with additional content
const randomFileUpdate = async (filepaths: fs.PathLike[], index = getRandomInt(filepaths.length)) => {
    const filepath = path.resolve(filepaths[index].toString());
    const fileContent = await readFileAsync(filepath, { encoding: 'utf-8' });
    await writeFileAsync(filepath, fileContent + casual.text);
}

const getRandomInt = (max: number, min = 1) => Math.max(
    Math.floor(Math.random() * Math.floor(max)), min
);