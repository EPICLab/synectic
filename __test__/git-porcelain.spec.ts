/* eslint-disable jest/no-commented-out-tests */
import * as path from 'path';
import { homedir } from 'os';

import * as git from '../src/containers/git-porcelain';
import * as io from '../src/containers/io';
import type { MockInstance } from './__mocks__/mock-fs-promise';
import { mock, file } from './__mocks__/mock-fs-promise';

describe('git.currentBranch', () => {
  let mockedInstance: MockInstance;

  beforeAll(async () => {
    const instance = await mock({
      baseRepo: {
        '.git': {
          HEAD: 'ref: refs/heads/master\n',
          branches: {},
          config: `[core]
	repositoryformatversion = 0
	filemode = true
	bare = false
	logallrefupdates = true
	ignorecase = true
	precomposeunicode = true`,
          description: 'Unnamed repository; edit this file \'description\' to name the repository.',
          hooks: {},
          info: {
            exclude: `# git ls-files --others --exclude-from=.git/info/exclude
# Lines that start with '#' are comments.
# For a project mostly in C, the following would be a good set of
# exclude patterns (uncomment them if you want to use them):
# *.[oa]
# *~
.DS_Store`
          },
          objects: {
            info: {},
            pack: {}
          },
          refs: {
            heads: {
              master: 'f204b02baf1322ee079fe9768e9593509d683412\n',
              foo: 'f204b02baf1322ee079fe9768e9593509d683412\n',
            },
            tags: {}
          },
          worktrees: {
            foo: {
              HEAD: 'ref: refs/heads/foo\n',
              ORIG_HEAD: 'f204b02baf1322ee079fe9768e9593509d683412\n',
              commondir: '../..\n',
              gitdir: `${process.cwd()}/foo/.git` // gitdir is an absolute path, and mock-fs uses process.cwd() as the path base
            }
          }
        }
      },
      foo: {
        '.git': 'gitdir: baseRepo/.git/worktrees/foo\n',
        bar: file({
          content: 'file contents',
          mtime: new Date(1)
        }),
        yez: {
          tam: {
            'som.js': file({ content: 'other information' })
          }
        }
      },
      qux: {
        '.git': {
          HEAD: 'f204b02baf1322ee079fe9768e9593509d683412\n',
          objects: {},
          refs: {
            heads: {
              master: 'f204b02baf1322ee079fe9768e9593509d683412\n',
            }
          }
        }
      },
      baz: {
        'sample.txt': 'non-tracked file and directory'
      }
    });
    return mockedInstance = instance;
  });

  afterAll(() => mockedInstance.reset());
  afterEach(jest.clearAllMocks);

  it('currentBranch resolves to Git branch name on a tracked worktree', async () => {
    jest.spyOn(io, 'isDirectory').mockResolvedValue(true); // mock-fs struggles to mock async IO calls, like io.isDirectory()
    return expect(git.currentBranch({ dir: 'baseRepo/' })).resolves.toBe('master');
  });

  it('currentBranch resolves to Git branch name on a linked woktree', async () => {
    jest.spyOn(io, 'isDirectory').mockResolvedValue(false); // mock-fs struggles to mock async IO calls, like io.isDirectory()
    return expect(git.currentBranch({ dir: 'foo/' })).resolves.toBe('foo');
  });

  it('currentBranch resolves to undefined on a tracked directory with detached HEAD', async () => {
    jest.spyOn(io, 'isDirectory').mockResolvedValue(true); // mock-fs struggles to mock async IO calls, like io.isDirectory()
    await expect(git.currentBranch({ dir: 'qux/' })).resolves.toBeUndefined();
  });

  it('currentBranch fails with an error on an untracked directory', async () => {
    jest.spyOn(io, 'isDirectory').mockResolvedValue(false); // mock-fs struggles to mock async IO calls, like io.isDirectory()
    await expect(git.currentBranch({ dir: 'baz/' })).rejects.toThrow(/ENOENT/);
  });
});

describe('git.getConfig', () => {
  let mockedInstance: MockInstance;

  beforeAll(async () => {
    const instance = await mock({
      [path.join(homedir(), '.gitconfig')]: file({
        content: `[user]
  name = Sandy Updates
  email = supdate@oregonstate.edu
[core]
  editor = vim
  whitespace = fix,-indent-with-non-tab,trailing-space,cr-at-eol`,
      }),
      '.git/config': file({
        content: `[user]
  name = Bobby Tables
  email = bdrop@oregonstate.edu
[credential]
  helper = osxkeychain
[pull]
  rebase = true
[alias]
  last = log -1 HEAD`,
      }),
    });
    return mockedInstance = instance;
  });

  afterAll(() => mockedInstance.reset());

  it('getConfig resolves global git-config value', async () => {
    return expect(git.getConfig('user.name', true)).resolves.toStrictEqual({ scope: 'global', value: 'Sandy Updates' });
  });

  it('getConfig resolves local git-config value', async () => {
    return expect(git.getConfig('user.name')).resolves.toStrictEqual({ scope: 'local', value: 'Bobby Tables' });
  });

  it('getConfig checks both scopes for git-config value', async () => {
    return expect(git.getConfig('core.editor')).resolves.toStrictEqual({ scope: 'global', value: 'vim' });
  });

  it('getConfig resolves none when no git-config value is found', async () => {
    return expect(git.getConfig('commenter.name')).resolves.toStrictEqual({ scope: 'none' });
  });
});

describe('git.setConfig', () => {
  let mockedInstance: MockInstance;

  beforeEach(async () => {
    const instance = await mock({
      [path.join(homedir(), '.gitconfig')]: file({
        content: `[user]
  name = Sandy Updates
  email = supdate@oregonstate.edu
[core]
  editor = vim`,
      }),
      '.git/config': file({
        content: `[user]
  name = Bobby Tables
  email = bdrop@oregonstate.edu
[credential]
  helper = osxkeychain`,
      }),
    });
    return mockedInstance = instance;
  });

  afterAll(() => mockedInstance.reset());

  it('setConfig adds value to global git-config', async () => {
    const updated = await git.setConfig('global', 'user.username', 'supdate');
    // mock.restore(); // required to prevent snapshot rewriting because of file watcher race conditions in Jest
    expect(updated).toMatchSnapshot();
  });

  it('setConfig updates values to local git-config', async () => {
    const updated = await git.setConfig('local', 'user.name', 'John Smith');
    // mock.restore(); // required to prevent snapshot rewriting because of file watcher race conditions in Jest
    expect(updated).toMatchSnapshot();
  });

  it('setConfig deletes values from local git-config', async () => {
    const updated = await git.setConfig('local', 'user.name', undefined);
    // mock.restore(); // required to prevent snapshot rewriting because of file watcher race conditions in Jest
    expect(updated).toMatchSnapshot();
  });
});

describe('git.getRepoRoot', () => {
  let mockedInstance: MockInstance;

  beforeAll(async () => {
    const instance = await mock({
      foo: {
        '.git': 'gitdir: baseRepo/.git/worktrees/foo\n',
        bar: file({
          content: 'file contents',
          mtime: new Date(1)
        }),
        yez: {
          tam: {
            'som.js': file({ content: 'other information' })
          }
        }
      },
      qux: {
        '.git': {
          HEAD: 'f204b02baf1322ee079fe9768e9593509d683412\n',
          objects: {},
          refs: {
            heads: {
              master: 'f204b02baf1322ee079fe9768e9593509d683412\n',
            }
          }
        }
      },
      baz: {
        'sample.txt': 'non-tracked file and directory'
      }
    });
    return mockedInstance = instance;
  });

  afterAll(() => mockedInstance.reset());

  it('getRepoRoot resolves to Git root directory on file in tracked directory', async () => {
    return expect(git.getRepoRoot('foo/yez/tam/som.js')).resolves.toBe('foo');
  });

  it('getRepoRoot resolves to undefined on file in untracked directory', async () => {
    await expect(git.getRepoRoot('baz/')).resolves.toBeUndefined();
  });
});

describe('git.getStatus', () => {
  let mockedInstance: MockInstance;

  beforeAll(async () => {
    const instance = await mock({
      foo: {
        '.git': 'gitdir: baseRepo/.git/worktrees/foo\n',
        bar: file({
          content: 'file contents',
          mtime: new Date(1)
        }),
        yez: {
          tam: {
            'som.js': file({ content: 'other information' })
          }
        }, 
        pez: {
          '.git': {
            'HEAD': 'ref: refs/heads/feature/test',
            'config': '[core]\nrepositoryformatversion = 0\nfilemode = true\nbare = false\nlogallrefupdates = true\nignorecase = true\nprecomposeunicode = true\n[remote "origin"]\nurl = git@github.com:test/test.git\nfetch = +refs / heads/*:refs/remotes/origin/*\n[branch "master"]\nremote = origin\nmerge = refs/heads/master',
            objects: {
              'e2': {
                '7bb34b0807ebf1b91bb66a4c147430cde4f08f': Buffer.from([98, 108, 111, 98, 32, 50, 53, 0, 77, 121, 32, 100, 97, 116, 97, 32, 102, 105, 116, 115, 32, 111, 110, 32, 111, 110, 101, 32, 108, 105, 110, 101, 10]),
              },
              '42': {
                '2a8a27eebd3798c661f2c0788dc8d6dfe597a1': 'blob 26\x00My data fits on line line\n'
              }
            },
            refs: {
              heads: {
                feature: {
                  'test': '4c40253aace4ffa46c943311d77232cb5d4ffe93'
                },
                'master': '4c40253aace4ffa46c943311d77232cb5d4ffe93',
                'remote-only': 'a81c46a181052b4bbb0037b7ab192540c4234054'
              },
            }
          }
        }
      },
      qux: {
        '.git': {
          HEAD: 'f204b02baf1322ee079fe9768e9593509d683412\n',
          objects: {},
          refs: {
            heads: {
              master: 'f204b02baf1322ee079fe9768e9593509d683412\n',
            }
          }
        },
        'tracked-file.js': 'directory is tracked by git',
        'another-file.ts': 'directory is tracked by git, but the git repo is currently in a detached HEAD state'
      },
      baz: {
        'sample.txt': 'non-tracked file and directory'
      }
    });
    return mockedInstance = instance;
  });

  afterAll(() => mockedInstance.reset());

  // it('getStatus resolves Git status on tracked file', async () => {
  //   await expect(git.getStatus(path.resolve('qux/tracked-file.js'))).resolves.toBe('unmodified');
  // });

  // it('getStatus resolves Git status on tracked directory', async () => {
  //   await expect(git.getStatus(path.resolve('foo/pez'))).resolves.toBe('unmodified');
  // });

  it('getStatus resolves to undefined on untracked file', async () => {
    await expect(git.getStatus('baz/sample.txt')).resolves.toBeUndefined();
  });

  it('getStatus resolves to undefined on untracked directory', async () => {
    await expect(git.getStatus('foo/bar/')).resolves.toBeUndefined();
  });
});