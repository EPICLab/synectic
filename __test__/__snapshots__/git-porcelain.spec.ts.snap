// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`git.setConfig setConfig adds value to global git-config 1`] = `
"[user]
name = Sandy Updates
email = supdate@oregonstate.edu
username = supdate

[core]
editor = vim
"
`;

exports[`git.setConfig setConfig deletes values from local git-config 1`] = `
"[user]
email = bdrop@oregonstate.edu

[credential]
helper = osxkeychain
"
`;

exports[`git.setConfig setConfig updates values to local git-config 1`] = `
"[user]
name = John Smith
email = bdrop@oregonstate.edu

[credential]
helper = osxkeychain
"
`;
